src/jit/flowgraph.cpp

```diff
@@ -9769,49 +9766,40 @@ void                Compiler::fgCompactBlocks(BasicBlock* block, BasicBlock* bNe
     // control-flow choice was constant-folded away.  So probably phi's need to go away,
     // as well, in favor of one of the incoming branches.  Or at least be modified.
 
-    assert((block->IsLIR() == bNext->IsLIR()) || (block->IsLIR() && bNext->isEmpty()) || (block->isEmpty() && bNext->IsLIR()));
+    assert(block->IsLIR() == bNext->IsLIR());
     if (block->IsLIR() || bNext->IsLIR())
     {
-        LIR::Range blockRange = LIR::AsRange(block);
-        LIR::Range nextRange = LIR::AsRange(bNext);
-
-        GenTree* blockBegin = blockRange.Begin();
-        GenTree* blockFirstNonPhi = blockRange.FirstNonPhiNode();
-
-        GenTree* nextBegin = nextRange.Begin();
-        GenTree* nextFirstNonPhi = nextRange.FirstNonPhiNode();
-
-        GenTree* insertionPoint = nullptr;
-        if (blockBegin != blockFirstNonPhi)
-        {
-            insertionPoint = blockFirstNonPhi != blockRange.End() ? blockFirstNonPhi->gtPrev : blockRange.EndExclusive();
-        }
+        LIR::Range& blockRange = LIR::AsRange(block);
+        LIR::Range& nextRange = LIR::AsRange(bNext);
 
         // Does the next block have any phis?
-        if (nextBegin != nextFirstNonPhi)
+        GenTree* nextFirstNonPhi = nullptr;
+        LIR::ReadOnlyRange nextPhis = nextRange.PhiNodes();
+        if (!nextPhis.IsEmpty())
         {
-            GenTree* lastPhi = nextFirstNonPhi != nextRange.End() ? nextFirstNonPhi->gtPrev : nextRange.EndExclusive();
+            GenTree* blockFirstNonPhi = blockRange.FirstNonPhiNode();
+            nextFirstNonPhi = nextPhis.LastNode()->gtNext;
 
-            LIR::Range nextPhis = LIR::AsRange(nextBegin, lastPhi);
-            nextRange.Remove(nextPhis);
-
-            if (insertionPoint == nullptr)
+            LIR::Range phisToMove = nextRange.Remove(std::move(nextPhis));
+            if (blockFirstNonPhi == nullptr)
             {
-                blockRange.InsertAtBeginning(nextPhis);
+                blockRange.InsertAtEnd(std::move(phisToMove));
             }
             else
             {
-                blockRange.InsertAfter(nextPhis, insertionPoint);
+                blockRange.InsertBefore(std::move(phisToMove), blockFirstNonPhi);
```

Comment:
With LastPhiNode(), this all might be clearer as:

            GenTree* blockLastPhi = blockRange.LastPhiNode();
            ...
            if (blockLastPhi == nullptr)
            {
                blockRange.InsertAtBeginning(std::move(phisToMove));
            }
            else
            {
                blockRange.InsertAfter(std::move(phisToMove), blockLastPhi);
            }

Mostly because it eliminates all the "non", which (IMO) confuses the logic.